[feature/mailer-optimization] Add interfaces and Events

lx-wnk opened the pull request:
### 1. Why is this change necessary?
To catch and enhance mails 

### 2. What does this change do, exactly?
- Add events to MailService.php
- Add Interfaces for decoration

### 3. Describe each step to reproduce the issue or behaviour.
///

### 4. Please link to the relevant issues (if any).
///

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

jhofm commented:
oh, someone else fixed this. ^^
https://github.com/shopware/shopware/pull/400/files
i just lost a holiday because of that query. again.

bcremer commented:
@wesolowski @jhofm merged with commit https://github.com/shopware/shopware/commit/03a06be77eecfb26eb9563310e5291b17f1a7a24.

Thanks for your contribution :+1:

hlohaus commented:
Nice. But you can catch mails:
https://github.com/hlohaus/SheMailer/blob/master/src/Resources/config/services.xml#L8
:)
A question: Can i access then custom fields?
https://github.com/shopware/platform/pull/340

lx-wnk commented:
@hlohaus
That might be possible but if you want to handle it config based it would be easier to use the events. I think also that more events in that case would provide a better usability for enhancements.

I think you should be able to but honestly I have not tried it yet. Although maybe the events needs to be enhanced

hlohaus commented:
Hi, thanks for you answer and pull request.
if you hadn't made it, I would have made it.
Can you integrate my pull request in yours?
Then it possible to something like that: 

https://github.com/hlohaus/platform/commit/20285089150daa3dfdbda0f29da1510d6a4e76b2#diff-98867856d544c8cd22262d2d0b47489bR169

@shopwareBot What do you think?

lx-wnk commented:
@hlohaus you're welcome.
I'll integrate the PR as soon as it is tagged for or just merged - whatever happens first.

ssltg commented:
Hi Alex,
thank you for your contribution. We imported your PR and will merge it soon.
👍

hlohaus commented on src/Core/Content/MailTemplate/Service/Event/MailBeforeSendEvent.php:
```diff
@@ -0,0 +1,80 @@
+    public function getLogData(): array
+    {
+        return [
+            'subject' => $this->subject,
```
maybe wrong?

lx-wnk replied:
I made some changes, could you take a look again?

hlohaus commented on src/Core/Content/MailTemplate/Service/MailerFactoryInterface.php:
```diff
@@ -0,0 +1,8 @@
+
+interface MailerFactoryInterface
+{
+    public function create(\Swift_Message $message): \Swift_Mailer;
```
maybe wrong

hlohaus reviewed (changes requested):
additionally customFields can be integrated.

lx-wnk commented:
> additionally customFields can be integrated.

I think the current implementation will work for this too. If you want to enhance the message or parts of it you could just subscribe to the event and handle the custom fields there. If it is not explicit needed I think an implementation would be unnecessary.

hlohaus commented:
I tried it out. There was one bug and it was not possible to get the custom fields or the mail template.
But this works:
https://github.com/hlohaus/platform/commit/f8ca94c1be436f7b4fe9e3be4635a5c02e21e5bc

hlohaus commented on src/Core/Content/MailTemplate/Service/MailService.php:
```diff
@@ -74,8 +76,8 @@ class MailService
     public function __construct(
```
implements MailServiceInterface...

lx-wnk replied:
Is added

Commit:
[feature/mailer-optimization] ] Add interfaces and Events
- Add events to MailService
- Add customFields to databag

lx-wnk commented:
@hlohaus and @ssltg could you both please review my changes again?
I changed the target to 6.1 and added thre template id and the custom fields to the databag according to a discussion with @hlohaus

mitelg commented:
thanks for your contribution @jinnoflife :+1: :tada: :blue_heart:

hlohaus commented:
@shopwareBot please import https://github.com/shopware/platform/commit/af9f23bcb3ddde7c11a8849541842253bd4838b0
not 
https://github.com/shopware/core/commit/cd215853c5352f1193398aba75100b56814e6d25.
the second are broken.

mitelg commented:
hey @hlohaus 
we have reworked the PR a bit anyway. Would you mind having a look at the changes and see if they still fit your needs? Otherwise another PR as follow up would be nice :blush: :+1:

lx-wnk commented:
Hey @mitelg is there something missing or to do for @hlohaus and/or me?

I saw that the [6.2 RC is without](https://github.com/shopware/platform/blob/master/CHANGELOG-6.2.md) these changes.

**Edit:**
Never mind. Just saw that the 6.2-tag contains everything 🙂
